File: packages/titirilquen_core/src/titirilquen_core/demand/utility.py
"""Funciones de utilidad por modo — logit multinomial.

Portado de `titirilquen-repo/app.py:281-338`. Incluye la descomposición
(breakdown) de cada utilidad para facilitar la visualización educativa.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from titirilquen_core.city import CiudadLineal
from titirilquen_core.config import DemandConfig, StratumId

Modo = Literal["Auto", "Metro", "Bici", "Caminata"]

UTIL_IMPOSIBLE = -9999.0
"""Valor que representa utilidad `−∞` para modos fuera de dominio
(caminata > 30 min, bici > 45 min, auto sin tenencia)."""


@dataclass(frozen=True)
class TiemposObservados:
    """Tiempos observados desde la oferta para una celda específica.

    Si `None` para toda una simulación, se usan tiempos de flujo libre
    (iteración 0 del loop de equilibrio).
    """

    auto_total: float
    bici_total: float
    tren_acceso: float
    tren_espera: float
    tren_viaje: float


@dataclass(frozen=True)
class UtilityBreakdown:
    """Descomposición de la utilidad de un modo. Útil para visualización."""

    modo: Modo
    valor: float
    asc: float
    v_tiempo: float
    v_costo: float
    v_penalizaciones: float = 0.0
    feasible: bool = True


def _tiempos_flujo_libre(dist_km: float, globales) -> TiemposObservados:
    return TiemposObservados(
        auto_total=(dist_km / globales.v_auto) * 60,
        bici_total=(dist_km / globales.v_bici) * 60,
        tren_acceso=10.0,
        tren_espera=5.0,
        tren_viaje=(dist_km / globales.v_metro) * 60,
    )


def calcular_utilidades(
    *,
    estrato: StratumId,
    celda_origen: int,
    tiene_auto: bool,
    ciudad: CiudadLineal,
    config: DemandConfig,
    tiempos_observados: TiemposObservados | None,
) -> dict[Modo, UtilityBreakdown]:
    """Calcula la utilidad de cada modo para un agente en una celda dada.

    Si `tiempos_observados is None` (iteración 0), se asumen tiempos de flujo libre.
    """
    betas = config.estratos[estrato].betas
    penal = betas.penalizaciones_fisicas
    gl = config.globales

    dist_km = abs(ciudad.cbd_index - celda_origen) * ciudad.ancho_celda_km

    tiempos = tiempos_observados if tiempos_observados is not None else _tiempos_flujo_libre(dist_km, gl)
    t_cam = (dist_km / gl.v_caminata) * 60

    # AUTO
    c_auto = dist_km * gl.costo_combustible_km + gl.costo_parking
    asc_auto = betas.asc_auto
    v_t_auto = betas.b_tiempo_viaje * tiempos.auto_total
    v_c_auto = betas.b_costo * c_auto
    if tiene_auto:
        v_auto = asc_auto + v_t_auto + v_c_auto
        auto_breakdown = UtilityBreakdown("Auto", v_auto, asc_auto, v_t_auto, v_c_auto, feasible=True)
    else:
        auto_breakdown = UtilityBreakdown("Auto", UTIL_IMPOSIBLE, 0, 0, 0, feasible=False)

    # METRO
    c_metro = gl.costo_tarifa_metro
    asc_metro = betas.asc_metro
    v_t_metro = (
        betas.b_tiempo_viaje * tiempos.tren_viaje
        + betas.b_tiempo_espera * tiempos.tren_espera
        + betas.b_tiempo_caminata * tiempos.tren_acceso
    )
    v_c_metro = betas.b_costo * c_metro
    v_metro = asc_metro + v_t_metro + v_c_metro
    metro_breakdown = UtilityBreakdown("Metro", v_metro, asc_metro, v_t_metro, v_c_metro, feasible=True)

    # BICI — penalizaciones aditivas escalonadas (ver D-02)
    if tiempos.bici_total > 45:
        bici_breakdown = UtilityBreakdown("Bici", UTIL_IMPOSIBLE, 0, 0, 0, feasible=False)
    else:
        p = 0.0
        if tiempos.bici_total > 10:
            p += penal.bici_10
        if tiempos.bici_total > 20:
            p += penal.bici_20
        if tiempos.bici_total > 30:
            p += penal.bici_30
        asc_bici = betas.asc_bici
        v_t_bici = betas.b_tiempo_viaje * tiempos.bici_total
        v_bici = asc_bici + v_t_bici + p
        bici_breakdown = UtilityBreakdown("Bici", v_bici, asc_bici, v_t_bici, 0.0, v_penalizaciones=p, feasible=True)

    # CAMINATA — usa b_tiempo_caminata (no b_tiempo_viaje); ver D-03
    if t_cam > 30:
        cam_breakdown = UtilityBreakdown("Caminata", UTIL_IMPOSIBLE, 0, 0, 0, feasible=False)
    else:
        p = 0.0
        if t_cam > 5:
            p += penal.walk_5
        if t_cam > 15:
            p += penal.walk_15
        if t_cam > 25:
            p += penal.walk_25
        asc_cam = betas.asc_caminata
        v_t_cam = betas.b_tiempo_caminata * t_cam
        v_cam = asc_cam + v_t_cam + p
        cam_breakdown = UtilityBreakdown("Caminata", v_cam, asc_cam, v_t_cam, 0.0, v_penalizaciones=p, feasible=True)

    return {
        "Auto": auto_breakdown,
        "Metro": metro_breakdown,
        "Bici": bici_breakdown,
        "Caminata": cam_breakdown,
    }

File: apps/web/src/lib/utility.ts
/**
 * Implementación TS de `titirilquen_core.demand.utility.calcular_utilidades`.
 * Mirror exacto del código Python para que el frontend pueda mostrar la
 * descomposición en tiempo real sin re-correr la simulación.
 */

import type { DemandConfig, Modo, StratumId } from "@/lib/types";

export interface TiemposObservados {
  auto_total: number;
  bici_total: number;
  tren_acceso: number;
  tren_espera: number;
  tren_viaje: number;
}

export interface UtilityBreakdown {
  modo: Modo;
  valor: number;
  asc: number;
  v_tiempo: number;
  v_costo: number;
  v_penalizaciones: number;
  feasible: boolean;
}

const UTIL_IMPOSIBLE = -9999;

function flujoLibreTiempos(distKm: number, g: DemandConfig["globales"]): TiemposObservados {
  return {
    auto_total: (distKm / g.v_auto) * 60,
    bici_total: (distKm / g.v_bici) * 60,
    tren_acceso: 10,
    tren_espera: 5,
    tren_viaje: (distKm / g.v_metro) * 60,
  };
}

export function calcularUtilidades({
  estrato,
  distKm,
  tieneAuto,
  config,
  tiempos,
}: {
  estrato: StratumId;
  distKm: number;
  tieneAuto: boolean;
  config: DemandConfig;
  tiempos?: TiemposObservados | null;
}): Record<Modo, UtilityBreakdown> {
  const stratum = config.estratos[estrato];
  if (!stratum) throw new Error(`Stratum ${estrato} not configured`);
  const betas = stratum.betas;
  const penal = betas.penalizaciones_fisicas;
  const gl = config.globales;

  const tt = tiempos ?? flujoLibreTiempos(distKm, gl);
  const tCam = (distKm / gl.v_caminata) * 60;

  // Auto
  const cAuto = distKm * gl.costo_combustible_km + gl.costo_parking;
  const vTAuto = betas.b_tiempo_viaje * tt.auto_total;
  const vCAuto = betas.b_costo * cAuto;
  const auto: UtilityBreakdown = tieneAuto
    ? {
        modo: "Auto",
        asc: betas.asc_auto,
        v_tiempo: vTAuto,
        v_costo: vCAuto,
        v_penalizaciones: 0,
        valor: betas.asc_auto + vTAuto + vCAuto,
        feasible: true,
      }
    : {
        modo: "Auto",
        asc: 0,
        v_tiempo: 0,
        v_costo: 0,
        v_penalizaciones: 0,
        valor: UTIL_IMPOSIBLE,
        feasible: false,
      };

  // Metro
  const cMetro = gl.costo_tarifa_metro;
  const vTMetro =
    betas.b_tiempo_viaje * tt.tren_viaje +
    betas.b_tiempo_espera * tt.tren_espera +
    betas.b_tiempo_caminata * tt.tren_acceso;
  const vCMetro = betas.b_costo * cMetro;
  const metro: UtilityBreakdown = {
    modo: "Metro",
    asc: betas.asc_metro,
    v_tiempo: vTMetro,
    v_costo: vCMetro,
    v_penalizaciones: 0,
    valor: betas.asc_metro + vTMetro + vCMetro,
    feasible: true,
  };

  // Bici
  let bici: UtilityBreakdown;
  if (tt.bici_total > 45) {
    bici = {
      modo: "Bici",
      asc: 0,
      v_tiempo: 0,
      v_costo: 0,
      v_penalizaciones: 0,
      valor: UTIL_IMPOSIBLE,
      feasible: false,
    };
  } else {
    let p = 0;
    if (tt.bici_total > 10) p += penal.bici_10;
    if (tt.bici_total > 20) p += penal.bici_20;
    if (tt.bici_total > 30) p += penal.bici_30;
    const vT = betas.b_tiempo_viaje * tt.bici_total;
    bici = {
      modo: "Bici",
      asc: betas.asc_bici,
      v_tiempo: vT,
      v_costo: 0,
      v_penalizaciones: p,
      valor: betas.asc_bici + vT + p,
      feasible: true,
    };
  }

  // Caminata
  let cam: UtilityBreakdown;
  if (tCam > 30) {
    cam = {
      modo: "Caminata",
      asc: 0,
      v_tiempo: 0,
      v_costo: 0,
      v_penalizaciones: 0,
      valor: UTIL_IMPOSIBLE,
      feasible: false,
    };
  } else {
    let p = 0;
    if (tCam > 5) p += penal.walk_5;
    if (tCam > 15) p += penal.walk_15;
    if (tCam > 25) p += penal.walk_25;
    const vT = betas.b_tiempo_caminata * tCam;
    cam = {
      modo: "Caminata",
      asc: betas.asc_caminata,
      v_tiempo: vT,
      v_costo: 0,
      v_penalizaciones: p,
      valor: betas.asc_caminata + vT + p,
      feasible: true,
    };
  }

  return { Auto: auto, Metro: metro, Bici: bici, Caminata: cam, Teletrabajo: {
    modo: "Teletrabajo",
    asc: 0, v_tiempo: 0, v_costo: 0, v_penalizaciones: 0, valor: 0, feasible: false,
  } };
}

export function probabilidadesLogit(utils: Record<Modo, UtilityBreakdown>): Record<Modo, number> {
  const feasibles = Object.values(utils).filter((u) => u.feasible);
  if (feasibles.length === 0) {
    return {
      Auto: 0, Metro: 0, Bici: 0, Caminata: 0, Teletrabajo: 0,
    } as Record<Modo, number>;
  }
  const maxV = Math.max(...feasibles.map((u) => u.valor));
  const exps = feasibles.map((u) => Math.exp(u.valor - maxV));
  const sum = exps.reduce((a, b) => a + b, 0);
  const out: Record<Modo, number> = {
    Auto: 0, Metro: 0, Bici: 0, Caminata: 0, Teletrabajo: 0,
  };
  feasibles.forEach((u, i) => {
    out[u.modo] = (exps[i] ?? 0) / sum;
  });
  return out;
}
